File: tests/lognormal_bias.R


# Simulate data
log_sigma_re <- 2
log_sigma <- 3
n_group <- 20
meanR <- 5

# Simulate random effects
re <- c()
x <- c()
group <- c()
ind <- 1
for(i in 1:n_group){
  re[i] <- rnorm(1, 0, exp(log_sigma_re))
  n_obs <- round(runif(1, min = 5, max = 100))
  # Simulate observations given random effects
  x <- c(x, rlnorm(n_obs, meanR + re[i], exp(log_sigma)))
  group <- c(group, rep(i-1, n_obs))
}

# Setup TMB
library(TMB)
library(TMBhelper)
version <- "lognormal_bias"
cpp_directory <- "tests"
cpp_file <- paste0(cpp_directory, "/", version)
TMB::compile(paste0(cpp_file, ".cpp"))
dyn.load(TMB::dynlib(paste0(cpp_file)))
params <- list(log_re = rep(0, n_group), log_sigma = 2, log_sigma_re = 2, log_meanR = 50)

# Mod 1 - not in like
datalist <- list(x = x, group = group, n_group = n_group, mod = 1)


# Estimate mean, deviations, and variance as fixed effects
obj = TMB::MakeADFun(datalist, parameters = params,  DLL = version, silent = TRUE, random = "log_re")
opt = TMBhelper::Optimize( obj )
rep <- sdreport(obj)

File: tests/lognormal_bias.cpp
#include <TMB.hpp>
template<class Type>
Type objective_function<Type>::operator() (){
  // Set-up
  DATA_VECTOR(x);
  DATA_IVECTOR(group);
  DATA_INTEGER(n_group);
  DATA_INTEGER( mod );

  PARAMETER_VECTOR(log_re);
  PARAMETER(log_sigma_re);
  PARAMETER(log_sigma);
  PARAMETER(log_meanR);

  // Model objects
  Type sigma_re = exp(log_sigma_re);
  Type sigma = exp(log_sigma);
  Type jnll = 0;

  // Random effects likelihood
  for(int j = 0; j < n_group; j++){
    if(mod == 1){
      jnll -= dnorm(log_re[j], Type(0), sigma_re, true);
    }
    if(mod == 2){
      jnll -= dnorm(log_re[j] - (sigma_re * sigma_re)/2, Type(0), sigma_re, true);
    }

  }

  vector<Type> rec(group.size());
  for(int i = 0; i < x.size(); i++){
    if(mod == 1){
      rec[i] = exp(log_meanR + log_re[group[i]] - (sigma_re * sigma_re) / 2);
    }

    if(mod == 2){
      rec[i] = exp(log_meanR + log_re[group[i]]);
    }
  }

  // Likelihood
  for(int i = 0; i < x.size(); i++){
    jnll -= dnorm( x[i], rec[i], sigma, true );
  }

  REPORT(sigma_re);
  REPORT(sigma);
  REPORT(log_meanR);
  REPORT(rec);

  ADREPORT(rec);
  ADREPORT(sigma);
  ADREPORT(sigma_re);

  return jnll;
}
